doc/source/rllib/metrics-logger.rst

```diff
@@ -0,0 +1,431 @@
+
+    logger.reduce()
+    logger.peek("some_items")  # expect an empty list: []
+
+
+Logging a set of nested scalar values
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In case you are logging a nested structure of values, for example
+``{"time_s": 0.1, "lives": 5, "rounds_played": {"player1": 10, "player2": 4}}`` and all values have the exact same log settings
+in terms of the ``reduce``, ``clear_on_reduce``, ``window``, etc arguments, you can also call the shortcut
+:py:meth:`~ray.rllib.utils.metrics.metrics_logger.MetricsLogger.log_dict` method to do so:
+
+
+.. testcode::
+
+
+
+    from ray.rllib.utils.metrics.metrics_logger import MetricsLogger
+
+    logger = MetricsLogger()
+
+    # Log a bunch of scalar values within a nested dict.
+    stats = {"player1": 100.0, "player2": 105.0}
+    logger.log_dict(stats, key="mean_scores", reduce="mean", window=10)
+
+    # Later, do the same again.
+    stats = {"player1": 150.0, "player2": 110.0}
+    logger.log_dict(stats, key="mean_scores")
+
+    print(logger.peek(("mean_scores", "player1")))  # <- expect 125.0
+
+
+Logging non-scalar data
+~~~~~~~~~~~~~~~~~~~~~~~
+
+:py:class:`~ray.rllib.utils.metrics_logger.MetricsLogger` isn't limited to scalar values.
+You can also use it to log images, videos, or any other complex data.
+
+Normally, you would chose the previously described ``reduce=None`` argument. For example, to
+log three consecutive image frames from a ``CartPole`` environment, do the following:
+
+.. testcode::
+
+    import gymnasium as gym
+
+    env = gym.make("CartPole-v1")
+
+    # Log three consecutive render frames from the env.
+    # Make sure to set ``clear_on_reduce=True`` to avoid memory leaks.
+    env.reset()
+    logger.log_value("some_images", value=env.render(), reduce=None, clear_on_reduce=True)
+    env.step(0)
+    logger.log_value("some_images", value=env.render())
+    env.step(1)
+    logger.log_value("some_images", value=env.render())
+
+
+
+
+Timing things
+~~~~~~~~~~~~~
+
+:py:class:`~ray.rllib.utils.metrics_logger.MetricsLogger` is context capable and offers the following
+simple API to log timer results.
+Notice that you can now time all your code blocks of interest inside your custom code through a single ``with-`` line:
+
+.. testcode::
+
+    import time
+    from ray.rllib.utils.metrics.metrics_logger import MetricsLogger
+
+    logger = MetricsLogger()
+
+    # First delta measurement:
+    with logger.log_time("my_block_to_be_timed", reduce="mean", ema_coeff=0.1):
+        time.sleep(1.0)
+
+    # EMA should be ~1sec.
+    assert 1.1 > logger.peek("my_block_to_be_timed") > 0.9
+
+    # Second delta measurement:
+    with logger.log_time("my_block_to_be_timed"):
+        time.sleep(2.0)
+
+    # EMA should be ~1.1sec.
+    assert 1.15 > logger.peek("my_block_to_be_timed") > 1.05
+
+Also note that the default logging behavior is through exponential mean averaging (EMA), with a default coefficient of 0.01.
```

Comment:
Maybe describing how to use a simple mean instead?

Reply:
Actually, we don't support this yet :D 
You can do a simple mean over a window, yes, but not over all lifetime, b/c we don't have a proper mechanism yet to  keep the list of values small. You could do `logger.log_value([key], [value], reduce="mean", window=float("inf"))`, but that would cause a mem leak at some point. We need to fix this by keeping track of the number of items logged thus far and then keeping the internal list smaller, but still providing the proper lifetime average. To be completely honest, not sure how useful a lifetime average would be (as opposed to an EMA, which makes more sense to report). But we should fix this either way ...